src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -251,12 +251,32 @@ export async function run(
   // A cache to hold built shader pipelines.
   const pipelineCache = new Map<String, GPUComputePipeline>();
 
-  // Submit all the cases in batches, each in a separate error scope.
-  const checkResults: Array<Promise<void>> = [];
+  // Submit all the cases in batches, rate-limiting to ensure not too many
+  // batches are in flight simultaneously.
+  const maxBatchesInFlight = 5;
+  let batchesInFlight = 0;
+  let resolvePromiseBlockingBatch: (() => void) | undefined = undefined;
+  const batchFinishedCallback = () => {
+    batchesInFlight -= 1;
+    // If there is any batch waiting on a previous batch to finish,
+    // unblock it now, and clear the resolve callback.
+    if (resolvePromiseBlockingBatch) {
+      resolvePromiseBlockingBatch();
+      resolvePromiseBlockingBatch = undefined;
+    }
+  };
+
   for (let i = 0; i < cases.length; i += casesPerBatch) {
     const batchCases = cases.slice(i, Math.min(i + casesPerBatch, cases.length));
 
-    t.device.pushErrorScope('validation');
```

Comment:
error scope deemed not necessary? (seems fine to me)
You know, it occurs to me... #2610

Reply:
Seems like it's intended to make the checks occur in a delayed manner, but it shouldn't ultimately be necessary.